### DISCUSSION TITLE
Integrate EJBCA as a Certificate Authority in Workspace ONE UEM

### DISCUSSION THREAD

**Discussion by FBL-dev** [Keyfactor/ejbca-ce#638] [Q&A]
Hello,

I am trying to integrate EJBCA into Workspace ONE UEM.
I have added the user certificate in the Windows Certificate Store of my Workspace ONE UEM/Airwatch Cloud.
On Workspace ONE, I indicate :
Authority Type: EJBCA
Server URL: https://my-pki-server/ejbca/ejbcaws/ejbcaws
Certificate: My_user.p12

However, when I test the connection, I get “Test is unsuccessful”.
And here's the error I get in the logs:

AirWatch.CloudConnector.CertificateService.CertificateService_EJBCA testConnectionRequest.AuthenticationCertificate/AuthorityURL cannot be null or empty.
Method: AirWatch.CloudConnector.CertificateService.CertificateService_EJBCA;  

Do you have any idea where my problem comes from?

Thanks

**Comment by lucas-carneiro** [comment]
Hi @FBL-dev, did you found the cause of it? Having the same issue

**Reply by FBL-dev** [reply]
Hi @lucas-carneiro,
No, unfortunately I'm still looking for a solution.
If you find a solution, please let me know.

**Comment by MB31** [comment]
Check if the ACC trusts the certificate from the workspace one uem cloud (GoDaddy).
I think the certificate used talking to EJBCA (RA), should be in de LocalMachine store, not in CurrentUser.

**Comment by smgehrig** [comment]
Do you have the AIA field configured in your certificate? http://www.pkiglobe.org/auth_info_access.html
This fixed it for us

**Reply by FBL-dev** [reply]
Hello,
Yes, I have the AIA field in the certificate

**Comment by lunor1** [comment]
@FBL-dev Have you found a solution? We have the same issue here.

**Reply by FBL-dev** [reply]
I'm still stuck on this issue as well. No solution on my end yet.

**Comment by primetomas** [comment]
I guess this is the documentation you are following?
https://docs.vmware.com/en/VMware-Workspace-ONE-UEM/2212/Certificate_Authority_Integrations/GUID-EJBCA.html

**Reply by lunor1** [reply]
Correct. And I have alos gone thru the blog: https://veucaddict.com/blog/integrate-ejbca-certificate-authority-in-workspace-one-uem/
After some troubleshooting (going direct to the VA-node instead of going thru a LB and created a new certificate for the RA User and imported it to UEM and Airwatch Cloud Connector Cert store) we are now getting another error mesage: **Not authorized to Web Services**.

We have applied most of the permissions to the user we use. Any idea?

**Reply by FBL-dev** [reply]
I don't fully understand your message:
“going direct to the VA-node instead of going thru a LB and created a new certificate for the RA User”.
Can you explain in more detail how you did this? 
And how did you get the certificate in .pfx format (mine is currently in .p12 format)?
Thank you

**Reply by lunor1** [reply]
We have a two node cluster. And first, we terminated the SSL in the LB. And that broke the Certificate auth from UEM to EJBCA.
So now we are going directly to the EJBCA.

I get the Certificate in pfx-format from our PKI team. Have you imported the certificate to the ACC local certificate store (Local Machine\Personal)?

Have you enabled Debug logs in ACC?

**Comment by primetomas** [comment]
If you get "Not authorized to Web Services" it sounds as if calls to EJBCA has been made. There should then be log items in server.log indicating if there are authorization issues.

**Reply by lunor1** [reply]
We found out that we did not had right permissions on the user we used. We can now issue certificates. 

@FBL-dev I can add for you that we added the certificate as .pfx in the configuration in UEM. And imported it to the local computer certificate store on the ACC.

**Reply by primetomas** [reply]
Fantastic. Is there something you think we can add to the documentation page (except correct links to the VMWare documentation that I already updated for the next release) to make it easier to get it working?
Any ideas and I can add it to the next doc update.

I.e. what can we add here to help: https://docs.keyfactor.com/ejbca/latest/vmware-workspace-one-uem-powered-by-airwatch

**Reply by lunor1** [reply]
I think you (and Omnissa) should add what URL that is needed e.g. https://my-pki-server/ejbca/ejbcaws/ejbcaws and that the the RA User certificate used in the integration should be in .pfx fomat and must be imported to the local machine certificate store on Airwatch Cloud Connector. 

I guess you should add what rights that is needed for the RA user as well, but im not sure exaclty what we added.

**Reply by primetomas** [reply]
Thanks. I updated the page, see attached screenshot. Hopefully it helps future users.
[Screenshot from 2024-11-13 15-07-16](https://github.com/user-attachments/assets/19a0b98e-b89c-4c28-8c8a-fd5b7aea41d9)

**Reply by lunor1** [reply]
Great! That will be helpful for future users. I have asked VMware/Omnissa to update thier documentation as well.

**Reply by primetomas** [reply]
I made an update to mention the TLS terminiation as well. 

EJBCA need to be able to use the TLS connection client certificate authentication. If you have a load balancer, consider that the TLS authentication must be visible to EJBCA. The easiest (to ensure it's working) is to terminate TLS in EJBCA, after this you can play with load balancer configuration.

**Reply by lunor1** [reply]
Fantastic!

**Reply by lunor1** [reply]
Next issue we have is that we cant send CSR requests from UEM to EJBCA that includes SAN attributes (Email, UPN and DNS-Name). In UEM we only get error messgae "Fault occurred while processing".

We have configured the EndEntityProfile to include Mail, DNS Name and MS UPN. We have also configured
```
If the Certificate Template to be mapped with the Certificate Profile uses "Supply in the request" option, following options need to be enabled in the Certificate Profile:
- Allow Subject DN Override by CSR
- Allow Extension Override for 2.5.29.17 (Subject Alt Name)
```
Any ideas?

**Reply by primetomas** [reply]
Check the server log it should be clear there what happens, i.e. if the correct profiles are being used. Sounds like a profile validation error, which should be clearly visible in the log.

**Reply by lunor1** [reply]
We found the issue. We have space in our CA name in EJBCA. When we requested certificates from Workspace ONE UEM that only have Subject and not SAN attributes, UEM were sending the correct CA name in the request. 

But as soon as we added SAN Attributes to the certificate request Workspace ONE UEM removed the space from the CA name and EJBCA gave a "CA not found" error, beacuse Worksapce ONE UEM sending wrong name. 

As a workaround we removed the spaces from the CA name and then it worked fine.

**Reply by primetomas** [reply]
Thanks for the information. I'll update the docs with this information.
